src/lib/operators/join_index.cpp

```diff
@@ -319,6 +367,59 @@ void JoinIndex::_append_matches(const BaseIndex::Iterator& range_begin, const Ba
   }
 }
 
+void JoinIndex::_append_matches_dereferenced(const ChunkID& probe_chunk_id, const ChunkOffset& probe_chunk_offset,
+                                             const PosList& index_table_matches) {
+  for (const auto& index_side_row_id : index_table_matches) {
+    _probe_pos_list->emplace_back(RowID{probe_chunk_id, probe_chunk_offset});
+    _index_pos_list->emplace_back((index_side_row_id));
+    _index_pos_dereferenced.emplace_back(true);
+  }
+}
+
+void JoinIndex::_append_matches_non_inner(const bool is_semi_or_anti_join) {
+  // For Full Outer and Left Join we need to add all unmatched rows for the probe side
+  if (_mode == JoinMode::Left || _mode == JoinMode::FullOuter) {
+    for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count(); ++probe_chunk_id) {
+      for (ChunkOffset chunk_offset{0}; chunk_offset < _probe_matches[probe_chunk_id].size(); ++chunk_offset) {
+        if (!_probe_matches[probe_chunk_id][chunk_offset]) {
+          _probe_pos_list->emplace_back(RowID{probe_chunk_id, chunk_offset});
+          _index_pos_list->emplace_back(NULL_ROW_ID);
+        }
+      }
+    }
+  }
+
+  // For Full Outer and Right Join we need to add all unmatched rows for the index side.
+  if (_mode == JoinMode::FullOuter || _mode == JoinMode::Right) {
+    for (ChunkID chunk_id{0}; chunk_id < _index_matches.size(); ++chunk_id) {
+      for (ChunkOffset chunk_offset{0}; chunk_offset < _index_matches[chunk_id].size(); ++chunk_offset) {
+        if (!_index_matches[chunk_id][chunk_offset]) {
+          _index_pos_list->emplace_back(RowID{chunk_id, chunk_offset});
+          _probe_pos_list->emplace_back(NULL_ROW_ID);
+        }
+      }
+    }
+  }
+
+  _probe_pos_list->shrink_to_fit();
+  _index_pos_list->shrink_to_fit();
+
+  // Write PosLists for Semi/Anti Joins, which so far haven't written any results to the PosLists
+  // We use `_probe_matches` to determine whether a tuple from the probe side found a match.
+  if (is_semi_or_anti_join) {
+    const auto invert = _mode == JoinMode::AntiNullAsFalse || _mode == JoinMode::AntiNullAsTrue;
```

Comment:
Are the NULL modes handled/tested anywhere? I don't think it changed with this PR, so if they are not, feel free to make it a separate issue.

Reply:
All join modes are tested by using the join test runner. Setting `invert = false;` and executing the test suite again resulted in 52 failed tests. In these tests either AntiNullAsFalse or AntiNullAsTrue was used.